File: 백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.py
from collections import deque

start, end = list(map(int,input().split()))
queue = deque()
queue.append([start, 0])

visited = [float('inf')]*100001
visited[start] = 0

answerTime = float('inf')
answerCnt = 0
while len(queue):
  curPos, curTime = queue.popleft()
  if curTime > answerTime:
    break
  if curPos == end:
    answerTime = curTime
    answerCnt += 1
    continue

  if curPos*2<=100000 and curTime+1 <= visited[curPos*2]:
    visited[curPos*2] = curTime + 1
    queue.append([curPos*2, curTime+1])

  if curPos+1<=100000 and curTime+1 <= visited[curPos+1]:
    visited[curPos+1] = curTime + 1
    queue.append([curPos+1, curTime+1])
  
  if curPos-1>=0 and curTime+1 <= visited[curPos-1]:
    visited[curPos-1] = curTime + 1
    queue.append([curPos-1, curTime+1])

print(answerTime)
print(answerCnt)

File: 백준/Gold/12851. 숨바꼭질 2/숨바꼭질 2.js
const input = require("fs")
  .readFileSync("/dev/stdin", "utf8")
  .trim()
  .split("\n")
  .map((line) => line.replace(/\r/, ""));

class Node {
  constructor(data) {
    this.data = data;
    this.next = null;
  }
}

class Queue {
  constructor() {
    this.head = null;
    this.rear = null;
    this.length = 0;
  }

  enqueue(data) {
    const node = new Node(data);
    if (!this.head) {
      this.head = node;
    } else {
      this.rear.next = node;
    }
    this.rear = node;
    this.length++;
  }

  dequeue() {
    if (!this.head) {
      return false;
    }
    const data = this.head.data;
    this.head = this.head.next;
    this.length--;

    return data;
  }
}

const solution = (input) => {
  const [start, end] = input[0].split(" ").map((v) => Number(v));
  const visited = Array(100001).fill(Infinity);
  visited[start] = 0;

  const queue = new Queue();
  queue.enqueue([start, 0]);

  let minTime = Infinity;
  let cnt = 0;

  while (queue.length) {
    const [pos, curTime] = queue.dequeue();
    if (curTime > minTime) break;

    if (pos === end) {
      minTime = curTime;
      cnt += 1;
    }

    if (pos * 2 <= 100000 && curTime + 1 <= visited[pos * 2]) {
      queue.enqueue([pos * 2, curTime + 1]);
      visited[pos * 2] = curTime + 1;
    }

    if (pos + 1 <= 100000 && curTime + 1 <= visited[pos + 1]) {
      queue.enqueue([pos + 1, curTime + 1]);
      visited[pos + 1] = curTime + 1;
    }

    if (pos - 1 >= 0 && curTime + 1 <= visited[pos - 1]) {
      queue.enqueue([pos - 1, curTime + 1]);
      visited[pos - 1] = curTime + 1;
    }
  }

  console.log(minTime);
  console.log(cnt);
};

solution(input);
